ceph/monitors.go

```diff
@@ -290,63 +292,95 @@ type cephFeatureGroup struct {
-
-		return err
-	}
+	wg := &sync.WaitGroup{}
 
 	stats := &cephMonitorStats{}
-	if err := json.Unmarshal(buf, stats); err != nil {
-		return err
-	}
+	var statsErr error
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// Ceph usage
+		cmd := m.cephUsageCommand()
+		buf, _, err := m.conn.MonCommand(cmd)
+		if err != nil {
+			m.logger.WithError(err).WithField(
+				"args", string(cmd),
+			).Error("error executing mon command")
 
-	// Ceph time sync status
-	cmd = m.cephTimeSyncStatusCommand()
-	buf, _, err = m.conn.MonCommand(cmd)
-	if err != nil {
-		m.logger.WithError(err).WithField(
-			"args", string(cmd),
-		).Error("error executing mon command")
+			statsErr = err
+			return
+		}
 
-		return err
-	}
+		if err := json.Unmarshal(buf, stats); err != nil {
+			statsErr = err
+		}
+	}()
 
 	timeStats := &cephTimeSyncStatus{}
-	if err := json.Unmarshal(buf, timeStats); err != nil {
-		return err
-	}
+	var timeErr error
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// Ceph time sync status
+		cmd := m.cephTimeSyncStatusCommand()
+		buf, _, err := m.conn.MonCommand(cmd)
+		if err != nil {
+			m.logger.WithError(err).WithField(
+				"args", string(cmd),
+			).Error("error executing mon command")
 
-	// Ceph versions
-	cmd, _ = CephVersionsCmd()
-	buf, _, err = m.conn.MonCommand(cmd)
-	if err != nil {
-		m.logger.WithError(err).WithField(
-			"args", string(cmd),
-		).Error("error executing mon command")
+			timeErr = err
+			return
+		}
 
-		return err
-	}
+		if err := json.Unmarshal(buf, timeStats); err != nil {
+			timeErr = err
+		}
+	}()
 
-	versions, err := ParseCephVersions(buf)
-	if err != nil {
-		m.logger.WithError(err).Error("error parsing ceph versions command")
-	}
+	var versionsErr error
+	var versions map[string]map[string]float64
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
 
-	// Ceph features
-	cmd = m.cephFeaturesCommand()
-	buf, _, err = m.conn.MonCommand(cmd)
-	if err != nil {
-		m.logger.WithError(err).WithField(
-			"args", string(cmd),
-		).Error("error executing mon command")
+		// Ceph versions
+		cmd, _ := CephVersionsCmd()
+		buf, _, err := m.conn.MonCommand(cmd)
+		if err != nil {
+			m.logger.WithError(err).WithField(
+				"args", string(cmd),
+			).Error("error executing mon command")
 
-		return err
+			versionsErr = err
+			return
+		}
+
+		versions, err = ParseCephVersions(buf)
+		if err != nil {
+			m.logger.WithError(err).Error("error parsing ceph versions command")
+		}
+	}()
+
+	var featuresErr error
+	var buf []byte
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		// Ceph features
+		cmd := m.cephFeaturesCommand()
+		buf, _, featuresErr = m.conn.MonCommand(cmd)
+		if featuresErr != nil {
+			m.logger.WithError(featuresErr).WithField(
+				"args", string(cmd),
+			).Error("error executing mon command")
+		}
+	}()
+
+	wg.Wait()
+	if featuresErr != nil || versionsErr != nil || timeErr != nil || statsErr != nil {
+		return errors.New("errors occurred in monitors collector")
```

Comment:
I think what you might want here (and could be useful elsewhere in this PR) is https://pkg.go.dev/golang.org/x/sync/errgroup

Reply:
I forgot this existed but it's perfect, thank you.